proposals/bootstrapper-proposal.md

```diff
@@ -0,0 +1,253 @@
+kind: CustomResourceDefinition
+metadata:
+  name: member.kubeflow.org
+spec:
+  group: kubeflow.org
+  version: v1alpha1
+  scope: Namespaced
+  names:
+    plural: members
+    singular: members
+    kind: Member
+    shortNames: ["mbr"]
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: kubeflow:admin,
+rules:
+- apiGroups:
+  - *
+  resources:
+  - *
+  verbs:
+  - *
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: kubeflow:write,
+rules:
+- apiGroups:
+  - *
+  resources:
+  - configmaps,
+    pods,
+    services,
+    endpoints,
+    persistentvolumeclaims,
+    events
+  verbs:
+  - *
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: kubeflow:read,
+rules:
+- apiGroups:
+  - *
+  resources:
+  - configmaps,
+    pods,
+    services,
+    endpoints,
+    persistentvolumeclaims,
+    events
+  verbs:
+  - get,
+    watch,
+    list
+---
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: DecoratorController
+metadata:
+  name: kubeflow-bootstrapper
+spec:
+  resources:
+  - apiVersion: v1alpha1
+    resource: managednamespace
+    labelSelector:
+      matchExpressions:
+      - {key: namespace-created, operator: DoesNotExist}
+  hooks:
+    sync:
+      webhook:
+        url: http://kubeflow-bootstrapper.metacontroller/bootstrap
+---
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: CompositeController
+metadata:
+  name: kubeflow-deployer
+spec:
+  parentResource:
+    apiVersion: kubeflow.org/v1alpha1
+    resource: managednamespace
+  childResources:
+  - apiVersion: kubeflow.org/v1alpha1
+    resource: organization
+    updateStrategy:
+      method: InPlace
+  hooks:
+    sync:
+      webhook:
+        url: http://kubeflow-deployer.metacontroller/deploy
+```
+
+** 2. Deploy using kubectl create -f `<kubeflow>.yaml` **
+- Use a `<kubeflow>.yaml` which includes:
+  - PersistentVolume (from kubeflow_toolkit.yaml)<br/>
+  - ManagedNamespace CR (shown below)
+  - The metacontrollers noted in Section 1. above will create resources
+  - The metacontrollers hooks are still **WIP**
+
+``` yaml
+apiVersion: kubeflow.org/v1alpha1
+kind: ManagedNamespace
+metadata:
+  name: kubeflow-jdoe
+organization:
+  name: Acme
+  teams:
+  - name: ds-delta
+    members:
+    - member:
+      github: jdoe
+      slack: jdoe
+      firstName: Jane
+      lastName: Doe
+      affiliation: Doe Inc
+      email: jdoe@doeinc.com
+    - member:
+      github: msmith
+      slack: msmith
+      firstName: Mike
+      lastName: Smith
+      affiliation: Doe Inc
+      email: msmith@doeinc.com
+```
+
+
+
+## Alternatives Considered
+Using [kubeless](https://kubeless.io/) functions rather than [metacontroller](https://github.com/GoogleCloudPlatform/metacontroller) controllers
+
+Using the [Application CRD](https://github.com/kubernetes-sigs/application#what-is-the-application-crd) instead of the ManagedNamespace CRD.
```

Comment:
Why did you rule out Application CRD? I think the Application CRD solves different problems. Its basically a way of specifying that a group of K8s resources together form an application. I don't think it handles deployment or lifecycle management.

Reply:
I thought it had some aspects that might allow it to be used or extended in a way that would address what this proposal is suggesting and could also be leveraged for generating and deploying different 'bundles' (pytorch, tf, ...)

Reply:
I think the initial version of the Application CRD is to just provide metadata. i.e. we could create an Application CRD and list all the differen K8s resources that are part of Kubeflow. I think currently application CRD is just intended to serve as a container to make it easy for things like UIs to understand which resources should be grouped together.

I think longer term it might be used for lifecycle management but I'm not sure yet.

Based on the other discussion, it seems like lifecycle management might be different from what you are proposing in this proposal.

Reply:
sure, i can open an issue on this independent of the bootstrapper modifications and lifecycle